drivers/storage/portworx/component/pxrepo.go

```diff
@@ -0,0 +1,203 @@
+func (p *pxrepo) IsEnabled(cluster *corev1.StorageCluster) bool {
+	return cluster.Spec.PxRepo != nil && cluster.Spec.PxRepo.Enabled
+}
+
+func (p *pxrepo) Reconcile(cluster *corev1.StorageCluster) error {
+	ownerRef := metav1.NewControllerRef(cluster, corev1.SchemeGroupVersion.WithKind("StorageCluster"))
+
+	if err := p.createPxRepoService(cluster.Namespace, ownerRef); err != nil {
+		return err
+	}
+
+	if err := p.createPxRepoDeployment(cluster, ownerRef); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (p *pxrepo) Delete(cluster *corev1.StorageCluster) error {
+	ownerRef := metav1.NewControllerRef(cluster, corev1.SchemeGroupVersion.WithKind("StorageCluster"))
+
+	if err := k8sutil.DeleteService(p.k8sClient, PxRepoServiceName, cluster.Namespace, *ownerRef); err != nil {
+		return err
+	}
+
+	if err := k8sutil.DeleteDeployment(p.k8sClient, PxRepoDeploymentName, cluster.Namespace, *ownerRef); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (p *pxrepo) MarkDeleted() {
+}
+
+func (p *pxrepo) createPxRepoService(
+	clusterNamespace string,
+	ownerRef *metav1.OwnerReference,
+) error {
+	service := v1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            PxRepoServiceName,
+			Namespace:       clusterNamespace,
+			OwnerReferences: []metav1.OwnerReference{*ownerRef},
+		},
+		Spec: v1.ServiceSpec{
+			Selector: getPxRepoDeploymentLabels(),
+			Ports: []v1.ServicePort{
+				{
+					Name:       "default",
+					Protocol:   v1.ProtocolTCP,
+					Port:       int32(pxRepoServicePort),
+					TargetPort: intstr.FromInt(pxRepoServicePort),
+				},
+			},
+			Type: v1.ServiceTypeClusterIP,
+		},
+	}
+
+	return k8sutil.CreateOrUpdateService(p.k8sClient, &service, ownerRef)
+}
+
+func getPxRepoDeploymentLabels() map[string]string {
+	return map[string]string{
+		"name": PxRepoDeploymentName,
+	}
+}
+
+func (p *pxrepo) createPxRepoDeployment(
+	cluster *corev1.StorageCluster,
+	ownerRef *metav1.OwnerReference,
+) error {
+	existingDeployment := &apps.Deployment{}
+	err := p.k8sClient.Get(
+		context.TODO(),
+		types.NamespacedName{
+			Name:      PxRepoDeploymentName,
+			Namespace: cluster.Namespace,
+		},
+		existingDeployment,
+	)
+	if err != nil && !errors.IsNotFound(err) {
+		return err
+	}
+
+	deployment := p.getPxRepoDeployment(cluster, ownerRef)
+
+	// If existingDeployment does not exist, modified will be true.
+	modified :=
+		k8sutil.GetImageFromDeployment(existingDeployment, pxRepoContainerName) != k8sutil.GetImageFromDeployment(deployment, pxRepoContainerName) ||
+			k8sutil.GetImagePullPolicyFromDeployment(existingDeployment, pxRepoContainerName) != k8sutil.GetImagePullPolicyFromDeployment(deployment, pxRepoContainerName) ||
+			util.HasPullSecretChanged(cluster, existingDeployment.Spec.Template.Spec.ImagePullSecrets) ||
+			util.HasNodeAffinityChanged(cluster, existingDeployment.Spec.Template.Spec.Affinity) ||
+			util.HaveTolerationsChanged(cluster, existingDeployment.Spec.Template.Spec.Tolerations)
+
+	if modified {
```

Comment:
As we discussed offline, we will need the 'isDeploymentCreated' flag for future changes to the deployment spec in operator. We may miss out on adding that in future, so it is better to add it now just like other deployment components.

Reply:
After looking at other components, some components do not have this flag. the flag is not needed anyway, i feel it's easier to understand without it.

Reply:
I believe all components that have a deployment or daemon set have the flag. It is not about being easier to understand, it is needed for future modifications to this deployment spec. For instance, in the next release of the operator if you add an env to the deployment in the code, it won't modify the existing deployment, if we don't add the flag.